### PR TITLE
[SWP-2082] Caching with singleton like

### PR DESCRIPTION
**Pull request by ersindemirtas-pp**
- Added a new property withAuthentication to make have access from parent class. withAuthentication makes sure that the access token is fetched and attached to request header. 
- Renamed confusing httpClient for apiClient. 
- Added laravel compatibility table to README.md
- Added phpcs
- Added a cache provider, to allow different cache drivers/adapters usable by this library. 
- Added Laravel Cache Adapter supports from Laravel 5+. 
- Added AbstractCacheableRequest to cache requests, any request can be cached by extending AbstractCacheableRequest.

### CODE REVIEW

File: src/Providers/Trustpilot/Request/AccessTokenRequest.php [hunk 1]
```diff
@@ -36,14 +36,23 @@ class AccessTokenRequest extends BaseRequest
...
-            self::CLIENT_ID,
-            self::CLIENT_SECRET,
-            self::USERNAME,
-            self::PASSWORD,
-        ];
+                self::CLIENT_ID,
+                self::CLIENT_SECRET,
+                self::USERNAME,
+                self::PASSWORD,
+               ];
```
Comment (AndreiaMatiasSilva) [inline src/Providers/Trustpilot/Request/AccessTokenRequest.php]: this indentation looks wrong

File: src/Request/AbstractCacheableRequest.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        $response = $this->apiClient->sendRequest(
+            $this->getRequest(),
+            $this->withAuthentication
+        );
+
+        $response->getStatusCode();
+
+        $responseBody = $this->apiClient->getResponseJson($response);
+
+        $this->cacheAdapter->set($this->getCacheableKey(), $responseBody, $this->cacheTtl);
```
Comment (AndreiaMatiasSilva) [inline src/Request/AbstractCacheableRequest.php]: the TTL in this case it's not from the response, ideally it should be, did we had issues when trying to achieve this?

File: tests/TestCase.php [hunk 3]
```diff
@@ -47,22 +50,32 @@ public function getMockedResponse($responseBody)
...
+     * @param array             $options
+     * @param array             $requiredFields
+     *
      * @return Mockery\Mock
      */
-    public function getMockedBaseRequest(AbstractApiClient  $client, array $options = [], $requiredFields = [])
-    {
-        $mock = Mockery::mock(BaseRequest::class, array($client, $options))
+    public function getMockedAbstractBaseRequest(
+        AbstractApiClient  $client,
```
Comment (AndreiaMatiasSilva) [inline tests/TestCase.php]: it looks like there's an extra space here

File: tests/TestCase.php [hunk 4]
```diff
@@ -47,22 +50,32 @@ public function getMockedResponse($responseBody)
...
     }
 
     /**
+     * Creates a mocked abstract base request.
+     *
      * @param AbstractApiClient $client
-     * @param array $options
-     * @param array $requiredFields
+     * @param array             $options
+     * @param array             $requiredFields
```
Comment (AndreiaMatiasSilva) [inline tests/TestCase.php]: this indentation was correct

File: src/Providers/Trustpilot/Request/AccessTokenRequest.php (hunk 1 above)
Comment (ersindemirtas-pp) [reply src/Providers/Trustpilot/Request/AccessTokenRequest.php]: its the phpcs autofix default indentation. Will revert.

File: tests/TestCase.php (hunk 4 above)
Comment (ersindemirtas-pp) [reply tests/TestCase.php]: phpcs auto indentation will revert.

File: src/Request/AbstractCacheableRequest.php (hunk 2 above)
Comment (ersindemirtas-pp) [reply src/Request/AbstractCacheableRequest.php]: I have updated it to use the expires_in from the response. This solution is just an idea but it works. Let me know what you think about it. I will add tests next for this.

File: src/Cache/CacheProvider.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * Class CacheProvider
+ */
+class CacheProvider
+{
+    private static $instance = null;
+
+    /**
+     * CacheFactory constructor.
+     */
+    private function __construct()
```
Comment (AndreiaMatiasSilva) [inline src/Cache/CacheProvider.php:22]: why is this private 🤔 ?

File: src/Cache/AbstractHasCacheTtlInResponse.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+use PodPoint\Reviews\Request\AbstractCacheableRequest;
+
+/**
+ * Class AbstractHasCacheTtlInResponse
+ */
+abstract class AbstractHasCacheTtlInResponse extends AbstractCacheableRequest
+{
+    /**
+     * The name of the field which has expires or ttl.
+     * @var
```
Comment (AndreiaMatiasSilva) [inline src/Cache/AbstractHasCacheTtlInResponse.php]: missing type

File: src/Cache/AbstractHasCacheTtlInResponse.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    protected $cacheTtlResponseField;
+
+    /**
+     * @param array|null $responseBody
```
Comment (AndreiaMatiasSilva) [inline src/Cache/AbstractHasCacheTtlInResponse.php]: I know it's probably a bit redundant but do you mind to add just a brief explanation

File: src/Cache/CacheProvider.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace PodPoint\Reviews\Cache;
+
+/**
+ * Class CacheProvider
+ */
+class CacheProvider
+{
+    private static $instance = null;
```
Comment (AndreiaMatiasSilva) [inline src/Cache/CacheProvider.php:17]: missing docblock

File: src/Providers/Trustpilot/Request/AccessTokenRequest.php [hunk 9]
```diff
@@ -58,11 +75,11 @@ public function getRequest(): Request
         $method = 'POST';
         $header = [
             'Authorization' => "Basic {$key}",
-            'Content-Type' => 'application/x-www-form-urlencoded'
+            'Content-Type'  => 'application/x-www-form-urlencoded',
```
Comment (AndreiaMatiasSilva) [inline src/Providers/Trustpilot/Request/AccessTokenRequest.php]: extra space not needed before `=>`

File: src/Providers/Trustpilot/Request/AccessTokenRequest.php [hunk 10]
```diff
@@ -58,11 +75,11 @@ public function getRequest(): Request
...
         ];
 
         $body = http_build_query([
-            'grant_type' => 'password',
+            'grant_type'   => 'password',
```
Comment (AndreiaMatiasSilva) [inline src/Providers/Trustpilot/Request/AccessTokenRequest.php]: extra space not needed before =>

File: src/Cache/CacheProvider.php (hunk 5 above)
Comment (ersindemirtas-pp) [reply src/Cache/CacheProvider.php:22]: The constructor is private to prevent initiation with outer code.

**Review by AndreiaMatiasSilva — COMMENTED** [review]
@ersindemirtas-pp we should also update the documentation for this right?

File: README.md [hunk 11]
```diff
@@ -87,13 +87,17 @@ $trustpilot->merchant()->getReviews((array) $serviceReviewsFilterOptions);
 
 ## Caching
 
-Using this library with Laravel and have it comptable a LaravelCacheAdapter has been added. To cache Request contents extend the request using the AbstractCacheableRequest class, this will automagicly cache the responses.
+Using this library with Laravel and to have it comptable, a LaravelCacheAdapter has been added. To cache Request contents extend the request using the AbstractCacheableRequest class, this will automagicly cache the responses.
```
Comment (AndreiaMatiasSilva) [inline README.md]: there's a typo `compatible `

File: README.md [hunk 12]
```diff
@@ -85,7 +85,18 @@ $trustpilot->merchant()->findReview((string) $reviewId);
 $trustpilot->merchant()->getReviews((array) $serviceReviewsFilterOptions);
 ```
 
-Compatibility table
+## Caching
+
+Using this library with Laravel and have it comptable a LaravelCacheAdapter has been added. To cache Request contents extend the request using the AbstractCacheableRequest class, this will automagicly cache the responses.
+
+The AbstractCacheableRequest has two optinal parameters $cacheTtl and $cacheKey, these can be overriden in demanded request. If no cacheKey is set the getCacheableKey is used to set the cache key and it will hash the class name using sha1 to have unique cacheKey.
```
Comment (AndreiaMatiasSilva) [inline README.md]: also here `optional`

File: README.md (hunk 11 above)
Comment (AndreiaMatiasSilva) [reply README.md]: also `automatically`

File: README.md (hunk 12 above)
Comment (AndreiaMatiasSilva) [reply README.md]: @ersindemirtas-pp can you please make sure the code has the correct style?

File: README.md [hunk 13]
```diff
@@ -85,6 +85,30 @@ $trustpilot->merchant()->findReview((string) $reviewId);
 $trustpilot->merchant()->getReviews((array) $serviceReviewsFilterOptions);
 ```
 
+## Caching
```
Comment (AndreiaMatiasSilva) [inline README.md]: I would add this part under the Development section. Is there anything that we should mention on Caching on the Usage section?

Comment (ersindemirtas-pp) [reply README.md]: @AndreiaMatiasSilva I have updated some of the parts of the code and included usage example.

File: README.md [hunk 14]
```diff
@@ -67,6 +67,13 @@ return [
 ];
 ```
 
+The Trustpilot AccessTokenRequest response will be cached using with the TTL of the access tokens expire time. The set cache driver you must implement the following before initialising ``PodPoint\Reviews\Reviews``.
```
Comment (AndreiaMatiasSilva) [inline README.md]: I would rephrase a bit to be:

The Trustpilot requires the access token to be cached, in order to ensure you do it when consuming this package ensure you need to set the cache driver.


```php
Example:
$cacheAdapter = new \PodPoint\Reviews\Cache\LaravelCacheAdapter();
\PodPoint\Reviews\Cache\CacheProvider::setInstance($cacheAdapter);
```

File: README.md [hunk 15]
```diff
@@ -96,6 +119,28 @@ Reviews PHP follows [semantic versioning](https://semver.org/) specifications.
 The MIT License (MIT). Please see [License File](https://github.com/Pod-Point/reviews-php/LICENCE) for more information.
 
 ## Development
+
+### Caching
+
+Using this library with Laravel and to have it compatible, a **LaravelCacheAdapter** has been added. To cache Requests response contents, extend the Request class using the **AbstractCacheableRequest** class, this will automatically cache the responses.
```
Comment (AndreiaMatiasSilva) [inline README.md]: not sure what you mean with this first sentence

Comment (ersindemirtas-pp) [reply README.md]: I have removed the first sentence, not required here since I have mentioned in the Laravel Caching. What I wanted to say is that to keep this library compatible with laravel a `LaravelCacheAdapter` has been added.

**Review by AndreiaMatiasSilva — DISMISSED** [review]

**Review by adteulade — DISMISSED** [review]
LGTM!

File: tests/Cache/AbstractHasCacheTtlInResponseTest.php [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $this->cacheAdapter = \Mockery::mock(CacheInterface::class);
+        CacheProvider::setInstance($this->cacheAdapter);
+
+        $this->mockedGuzzleRequest = \Mockery::mock(Request::class);
+        $this->httpClient = \Mockery::mock(Client::class, ['http_errors' => true])->makePartial();
+    }
+
+    /**
+     * Making sure that the expires_in access token
```
Comment (adteulade) [inline tests/Cache/AbstractHasCacheTtlInResponseTest.php]: Pedantic: `Making sure that the expires_in is converted in minutes and used as ttl for cache.`

**Review by adteulade — APPROVED** [review]
